adapters/solana/account_proof_geyser/src/lib.rs

```diff
@@ -0,0 +1,471 @@
+    };
+
+    let num_sigs = num_sigs.clone();
+    let parent_bankhash = Hash::from_str(&block.parent_bankhash).unwrap();
+    let blockhash = Hash::from_str(&block.blockhash).unwrap();
+
+    let mut account_hashes: Vec<(Pubkey, Hash)> = account_hashes_data
+        .iter()
+        .map(|(k, (_, v, _))| (k.clone(), v.clone()))
+        .collect();
+
+    let (inclusion, non_inclusion_left, non_inclusion_right, non_inclusion_inner) =
+        get_proof_pubkeys_required(&mut account_hashes, pubkeys_for_proofs);
+
+    let (non_inclusion_inner_adjacent_keys, non_inclusion_inner_mapping) =
+        get_keys_for_non_inclusion_inner(&non_inclusion_inner, &mut account_hashes);
+
+    let mut amended_leaves = inclusion.clone();
+
+    if non_inclusion_left.len() > 0 {
+        amended_leaves.push(account_hashes[0].0.clone());
+    }
+    if non_inclusion_right.len() > 0 {
+        amended_leaves.push(account_hashes[account_hashes.len() - 1].0.clone());
+    }
+    if non_inclusion_inner.len() > 0 {
+        amended_leaves.extend(non_inclusion_inner_adjacent_keys.iter().cloned());
+    }
+
+    let (accounts_delta_hash, account_proofs) =
+        calculate_root_and_proofs(&mut account_hashes, &amended_leaves);
+
+    let bank_hash = hashv(&[
+        parent_bankhash.as_ref(),
+        accounts_delta_hash.as_ref(),
+        &num_sigs.to_le_bytes(),
+        blockhash.as_ref(),
+    ]);
+
+    let proofs = assemble_account_delta_proof(
+        &account_hashes,
+        &account_hashes_data,
+        &account_proofs,
+        &inclusion,
+        &non_inclusion_left,
+        &non_inclusion_right,
+        &non_inclusion_inner,
+        &non_inclusion_inner_mapping,
+    )
+    .unwrap();
+
+    block_accumulator.remove(&slot);
+    processed_slot_account_accumulator.remove(&slot);
+    processed_transaction_accumulator.remove(&slot);
+
+    Ok(Update {
+        slot,
+        root: bank_hash,
+        proof: BankHashProof {
+            proofs,
+            num_sigs,
+            account_delta_root: accounts_delta_hash,
+            parent_bankhash,
+            blockhash,
+        },
+    })
+}
+
+fn handle_processed_slot(
+    slot: u64,
+    raw_slot_account_accumulator: &mut AccountHashAccumulator,
+    processed_slot_account_accumulator: &mut AccountHashAccumulator,
+    raw_transaction_accumulator: &mut TransactionSigAccumulator,
+    processed_transaction_accumulator: &mut TransactionSigAccumulator,
+) -> anyhow::Result<()> {
+    transfer_slot(
+        slot,
+        raw_slot_account_accumulator,
+        processed_slot_account_accumulator,
+    );
+    transfer_slot(
+        slot,
+        raw_transaction_accumulator,
+        processed_transaction_accumulator,
+    );
+    Ok(())
+}
+
+fn transfer_slot<V>(slot: u64, raw: &mut HashMap<u64, V>, processed: &mut HashMap<u64, V>) {
+    if let Some(entry) = raw.remove(&slot) {
+        processed.insert(slot, entry);
+    }
+}
+
+fn process_messages(
```

Comment:
Same comment here: This could use comments at each branch to explain what's going on

Reply:
added the necessary documentation!